File: Top_100_Liked_Questions/155._Min_Stack/solution.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.stack = [];
};

const last = arr => arr[arr.length - 1];

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    if (this.stack.length === 0) {
        this.stack.push([x,x]);
    } else {
        const currMin = last(this.stack)[1];
        this.stack.push([x, Math.min(x, currMin)]);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.stack.pop()
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return last(this.stack)[0];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return last(this.stack)[1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: Top_100_Liked_Questions/155._Min_Stack/solution.java
class MinStack {

    private List<Integer> stack;
    private List<Integer> minStack;

    /** initialize your data structure here. */
    public MinStack() {
        this.stack = new ArrayList<Integer>();
        this.minStack = new ArrayList<Integer>();
    }

    public void push(int x) {
        this.stack.add(x);

        if (this.minStack.size() == 0 || x <= this.minStack.get(this.minStack.size()-1)) {
            this.minStack.add(x);
        }
    }

    public void pop() {
        int popNum = this.stack.get(this.stack.size()-1);
        if (popNum == this.minStack.get(this.minStack.size()-1))
            this.minStack.remove(this.minStack.size()-1);

        this.stack.remove(this.stack.size()-1);
    }

    public int top() {
        return this.stack.get(this.stack.size()-1);
    }

    public int getMin() {
        return this.minStack.get(this.minStack.size()-1);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */